libslide/src/partial_evaluator/flatten.rs

```diff
@@ -0,0 +1,139 @@
+use crate::grammar::*;
+use crate::utils::{unflatten_binary_expr, UnflattenStrategy};
+
+use std::collections::{HashMap, VecDeque};
+use std::rc::Rc;
+
+pub fn flatten_expr(expr: &Rc<Expr>) -> Rc<Expr> {
+    match expr.as_ref() {
```

Comment:
why is expr getting match as_ref. shouldn't it already be ref due to the fact you're borrowing it.

Reply:
This gets a ref to the underlying `Expr`